File: sdk/typescript/src/client/execute-task.ts
/**
 * @centris/sdk - executeTask() High-Level API
 *
 * Run a natural-language task with optional structured output.
 * Cleaner surface than do() for programmatic use.
 */

import type { TSchema } from "@sinclair/typebox";
import type { Centris } from "./index.js";
import type { CentrisResult } from "./index.js";
import { extract } from "./extract.js";

/**
 * Options for executeTask().
 */
export interface ExecuteTaskOptions {
  /** Return immediately with taskId; poll via client.wait() */
  asyncMode?: boolean;
  /** When asyncMode, wait for completion (default: true) */
  wait?: boolean;
  /** Output schema for structured extraction (uses extract internally) */
  outputSchema?: TSchema;
  /** Additional context */
  context?: Record<string, unknown>;
}

/**
 * Result of executeTask().
 */
export interface ExecuteTaskResult<T = string> {
  /** Task output: text or structured data when outputSchema provided */
  output: T;
  /** Task ID */
  taskId: string;
  /** Task status */
  status: CentrisResult["status"];
  /** Actions performed */
  actions: CentrisResult["actions"];
  /** Usage metadata if available */
  usage?: Record<string, unknown>;
}

/**
 * Execute a natural-language task. High-level API over do().
 *
 * When outputSchema is provided, uses extract() internally to return typed data.
 * Otherwise returns the raw text result.
 *
 * @example
 * ```ts
 * const centris = new Centris();
 *
 * // Simple task
 * const r = await centris.executeTask("Open Gmail and read the first email subject");
 * console.log(r.output);
 *
 * // With structured output
 * const r2 = await centris.executeTask(
 *   "Search flights from NYC to LA, return top 3 options",
 *   { outputSchema: FlightSchema }
 * );
 * console.log(r2.output.flights);
 * ```
 */
export async function executeTask(
  client: Centris,
  task: string,
  options: ExecuteTaskOptions = {},
): Promise<ExecuteTaskResult<unknown>> {
  if (options.outputSchema) {
    const extractResult = await extract(client, task, options.outputSchema, {
      asyncMode: options.asyncMode,
      wait: options.wait ?? true,
      context: options.context,
    });
    return {
      output: extractResult.data,
      taskId: extractResult.taskId,
      status: extractResult.status,
      actions: extractResult.actions,
    };
  }

  const result = await client.do(task, {
    asyncMode: options.asyncMode ?? false,
    context: options.context ?? {},
  });

  let finalResult = result;
  if (
    options.wait !== false &&
    result.taskId &&
    (result.status === "queued" || result.status === "running")
  ) {
    finalResult = await client.wait(result.taskId);
  }

  return {
    output: finalResult.text,
    taskId: finalResult.taskId,
    status: finalResult.status,
    actions: finalResult.actions,
    usage: finalResult.usage,
  };
}

File: sdk/python/centris_sdk/execute_task.py
"""High-level executeTask API for Centris SDK."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, Dict, Optional

if TYPE_CHECKING:
    from centris_sdk.client import Centris


@dataclass
class ExecuteTaskResult:
    """Result of execute_task()."""

    output: Any
    task_id: str
    status: str
    actions: list
    usage: Optional[Dict[str, Any]] = None


def execute_task(
    client: "Centris",
    task: str,
    *,
    async_mode: bool = False,
    wait: bool = True,
    output_schema: Optional[Dict[str, Any]] = None,
    context: Optional[Dict[str, Any]] = None,
) -> ExecuteTaskResult:
    """Execute a natural-language task. High-level API over do().

    When output_schema is provided, uses extract() internally to return typed data.
    Otherwise returns the raw text result.

    Args:
        client: Centris client instance
        task: Natural language task description
        async_mode: If True, return immediately with task_id for polling
        wait: When async_mode, wait for completion (default True)
        output_schema: JSON Schema for structured output (uses extract internally)
        context: Additional context

    Returns:
        ExecuteTaskResult with output, task_id, status, actions

    Example:
        >>> result = centris.execute_task("Open Gmail and read first email subject")
        >>> print(result.output)

        >>> schema = {"type": "object", "properties": {"flights": {"type": "array"}}}
        >>> result = centris.execute_task("Search flights NYC to LA", output_schema=schema)
        >>> print(result.output["flights"])
    """
    if output_schema:
        from centris_sdk.extract import extract

        extract_result = extract(
            client,
            task,
            output_schema,
            async_mode=async_mode,
            wait=wait,
            context=context,
        )
        return ExecuteTaskResult(
            output=extract_result.data,
            task_id=extract_result.task_id,
            status=extract_result.status,
            actions=extract_result.actions,
        )

    result = client.do(task, async_mode=async_mode, context=context or {})

    if wait and result.task_id and result.status in ("queued", "running"):
        result = client.wait(result.task_id)

    return ExecuteTaskResult(
        output=result.text,
        task_id=result.task_id,
        status=result.status,
        actions=result.actions,
        usage=result.usage,
    )
